webapp/tests/Controller/ControllerRolesTest.php

```diff
@@ -0,0 +1,197 @@
+<?php declare(strict_types=1);
+
+namespace App\Tests\Controller;
+
+use App\Tests\BaseTest;
+
+class ControllerRolesTest extends BaseTest
+{
+    protected static $roles = [];
+
+    /**
+     * //See: https://www.oreilly.com/library/view/php-cookbook/1565926811/ch04s25.html
+     * Get all combinations of roles with at minimal the starting roles
+     * @return array $results
+     * @var string[] $possible_roles
+     * @var string[] $start_roles
+     */
+    protected function roleCombinations(array $start_roles, array $possible_roles)
+    {
+        // initialize by adding the empty set
+        $results = array($start_roles);
+
+        foreach ($possible_roles as $element) {
+            foreach ($results as $combination) {
+                array_push($results, array_merge(array($element), $combination));
+            }
+        }
+        return $results;
+    }
+
+    /**
+     * Crawl the webpage assume this is allowed and return all other links on the page
+     * @return string[] $urlsToCheck
+     * @var int $statusCode
+     * @var string $url
+     */
+    public function crawlPage(string $url, int $statusCode)
+    {
+        $crawler = $this->client->request('GET', $url);
+        $response = $this->client->getResponse();
+        $message = var_export($response, true);
+        $this->assertEquals($statusCode, $response->getStatusCode(), $message);
+        return array_unique($crawler->filter('a')->extract(['href']));
+    }
+
+    /**
+     * Test that having the role(s) gives access to all visible pages.
+     * This test should detect mistakes where a page is disabled when the user has a
+     * certain role instead of allowing when the correct role is there.
+     * @var string[] $combinations
+     * @var string[] $roleURLs
+     */
+    private function verifyAccess(array $combinations, array $roleURLs)
+    {
+        foreach ($combinations as static::$roles) {
+            foreach ($roleURLs as $url) {
+                // Documentation is not setup in the UnitTesting framework
+                if (substr($url, 0, 4) == '/doc') {
+                    continue;
+                } // The change-contest handles a different action
+                elseif (substr($url, 0, 21) == '/jury/change-contest/') {
+                    $this->crawlPage($url, 302);
+                } //Remove links to local page, external or application links
+                elseif ($url[0] != '#' && $url[0] != 'h' && $url != '/logout') {
+                    $this->crawlPage($url, 200);
+                }
+            }
+        }
+    }
+
+    /**
+     * Test that having the jury role is enough to view jury pages.
+     * This test should detect mistakes where a page is disabled when the user has a
+     * certain role instead of allowing when the correct role is there.
+     */
+    public function testJuryAccess()
+    {
+        $roleBaseUrl = '/jury';
```

Comment:
Would it be better if we combine these three functions (and the same for the last set) and use a `@dataProvider`? They contain quite a bit of duplication now.

Reply:
The reason I chose this was as how I understand UnitTesting you should have a test not do multiple tests. The current code already does more than one thing but having the multiple roles might really create one test which does all things.

I will look into the @dataProvider to see what it actually does. If its indeed this: https://symfony.com/doc/current/testing.html than I think your solution is much better than the current code.